# Phase 8B — Higher-Order Temporal Structure

## Exact three-step and conditional four-step sequence analysis

Phase 8B asks whether longer ordered displacement sequences contain
forward-versus-reversed information beyond net directional displacement.

The analysis design was pre-specified and committed before inspecting
Phase 8B outcome statistics.

Primary sequence length:

`3 steps`

Conditional secondary sequence length:

`4 steps`

The first data-dependent task is deliberately limited to a
**sample-support audit**.

At this stage we will determine only:

- the number of exact three-step sequences;
- the number of trajectory groups contributing to them;
- the number of exact four-step sequences;
- the number of trajectory groups contributing to them;
- whether the pre-specified four-step support gate is satisfied.

No classifier, AUC, p-value, null simulation, or irreversibility statistic
is computed in this section.

The pre-specified four-step gate is:

`at least 200 exact sequences AND at least 12 trajectory groups`.

In [1]:
import numpy as np
import pandas as pd

from _path import PROJECT_ROOT

from src.steps import validate_tracks
from src.sequences import build_k_step_sequences


CTC_REFERENCE_PATH = (
    PROJECT_ROOT
    / "data"
    / "MSC01_tracks_ctc_reference.csv"
)

TRACKMATE_RAW_PATH = (
    PROJECT_ROOT
    / "data"
    / "MSC01_tracks_clean.csv"
)

TRACKMATE_COMOVING_PATH = (
    PROJECT_ROOT
    / "data"
    / "MSC01_tracks_comoving.csv"
)


print("Project root:", PROJECT_ROOT)

print(
    "CTC reference exists:",
    CTC_REFERENCE_PATH.exists(),
)

print(
    "TrackMate raw exists:",
    TRACKMATE_RAW_PATH.exists(),
)

print(
    "TrackMate co-moving exists:",
    TRACKMATE_COMOVING_PATH.exists(),
)

Project root: C:\Users\Abolfazl.PH\Desktop\cell-irreversibility
CTC reference exists: True
TrackMate raw exists: True
TrackMate co-moving exists: True


In [2]:
ctc_reference_table = pd.read_csv(
    CTC_REFERENCE_PATH
)

trackmate_raw = pd.read_csv(
    TRACKMATE_RAW_PATH
)

trackmate_comoving = pd.read_csv(
    TRACKMATE_COMOVING_PATH
)


ctc_raw = (
    ctc_reference_table[
        [
            "track_id",
            "frame",
            "time_min",
            "x_um",
            "y_um",
        ]
    ]
    .rename(
        columns={
            "track_id": "cell_id",
            "time_min": "t_min",
        }
    )
    .sort_values(
        [
            "cell_id",
            "frame",
        ]
    )
    .reset_index(
        drop=True
    )
)


trackmate_raw = (
    trackmate_raw[
        [
            "cell_id",
            "frame",
            "t_min",
            "x_um",
            "y_um",
        ]
    ]
    .sort_values(
        [
            "cell_id",
            "frame",
        ]
    )
    .reset_index(
        drop=True
    )
)


trackmate_comoving = (
    trackmate_comoving[
        [
            "cell_id",
            "frame",
            "t_min",
            "x_um",
            "y_um",
        ]
    ]
    .sort_values(
        [
            "cell_id",
            "frame",
        ]
    )
    .reset_index(
        drop=True
    )
)


print(
    "CTC shape:",
    ctc_raw.shape,
)

print(
    "CTC tracks:",
    ctc_raw["cell_id"].nunique(),
)

print(
    "TrackMate raw shape:",
    trackmate_raw.shape,
)

print(
    "TrackMate raw tracks:",
    trackmate_raw["cell_id"].nunique(),
)

print(
    "TrackMate co-moving shape:",
    trackmate_comoving.shape,
)

print(
    "TrackMate co-moving tracks:",
    trackmate_comoving["cell_id"].nunique(),
)

CTC shape: (428, 5)
CTC tracks: 15
TrackMate raw shape: (474, 5)
TrackMate raw tracks: 35
TrackMate co-moving shape: (474, 5)
TrackMate co-moving tracks: 35


In [3]:
validate_tracks(
    ctc_raw
)

validate_tracks(
    trackmate_raw
)

validate_tracks(
    trackmate_comoving
)


trackmate_raw_structure = (
    trackmate_raw[
        [
            "cell_id",
            "frame",
            "t_min",
        ]
    ]
)

trackmate_comoving_structure = (
    trackmate_comoving[
        [
            "cell_id",
            "frame",
            "t_min",
        ]
    ]
)


trackmate_structure_identical = (
    trackmate_raw_structure.equals(
        trackmate_comoving_structure
    )
)


print(
    "TrackMate raw/co-moving structure identical:",
    trackmate_structure_identical,
)


assert trackmate_structure_identical

assert len(ctc_raw) == 428
assert ctc_raw["cell_id"].nunique() == 15

assert len(trackmate_raw) == 474
assert trackmate_raw["cell_id"].nunique() == 35


print(
    "Phase 8B trajectory structural validation: PASSED"
)

TrackMate raw/co-moving structure identical: True
Phase 8B trajectory structural validation: PASSED


In [4]:
ctc_three_step_raw = (
    build_k_step_sequences(
        ctc_raw,
        n_steps=3,
    )
)

ctc_four_step_raw = (
    build_k_step_sequences(
        ctc_raw,
        n_steps=4,
    )
)


trackmate_three_step_raw = (
    build_k_step_sequences(
        trackmate_raw,
        n_steps=3,
    )
)

trackmate_four_step_raw = (
    build_k_step_sequences(
        trackmate_raw,
        n_steps=4,
    )
)


print(
    "CTC exact 3-step sequences:",
    len(ctc_three_step_raw),
)

print(
    "CTC 3-step groups:",
    ctc_three_step_raw[
        "cell_id"
    ].nunique(),
)

print()

print(
    "CTC exact 4-step sequences:",
    len(ctc_four_step_raw),
)

print(
    "CTC 4-step groups:",
    ctc_four_step_raw[
        "cell_id"
    ].nunique(),
)

print()

print(
    "TrackMate exact 3-step sequences:",
    len(trackmate_three_step_raw),
)

print(
    "TrackMate 3-step groups:",
    trackmate_three_step_raw[
        "cell_id"
    ].nunique(),
)

print()

print(
    "TrackMate exact 4-step sequences:",
    len(trackmate_four_step_raw),
)

print(
    "TrackMate 4-step groups:",
    trackmate_four_step_raw[
        "cell_id"
    ].nunique(),
)

CTC exact 3-step sequences: 383
CTC 3-step groups: 14

CTC exact 4-step sequences: 369
CTC 4-step groups: 14

TrackMate exact 3-step sequences: 357
TrackMate 3-step groups: 28

TrackMate exact 4-step sequences: 329
TrackMate 4-step groups: 26


In [5]:
for (
    label,
    sequences,
    expected_steps,
) in [
    (
        "CTC 3-step",
        ctc_three_step_raw,
        3,
    ),
    (
        "CTC 4-step",
        ctc_four_step_raw,
        4,
    ),
    (
        "TrackMate 3-step",
        trackmate_three_step_raw,
        3,
    ),
    (
        "TrackMate 4-step",
        trackmate_four_step_raw,
        4,
    ),
]:

    assert (
        sequences[
            "sequence_span_frames"
        ]
        == expected_steps
    ).all()

    assert np.allclose(
        sequences[
            "sequence_span_min"
        ],
        expected_steps * 20.0,
    )

    assert (
        sequences[
            "frame_end"
        ]
        - sequences[
            "frame_start"
        ]
        == expected_steps
    ).all()


print(
    "Exact 3-step/4-step temporal-span validation: PASSED"
)

Exact 3-step/4-step temporal-span validation: PASSED


In [6]:
FOUR_STEP_MIN_SEQUENCES = 200
FOUR_STEP_MIN_GROUPS = 12


support_by_source = pd.DataFrame(
    [
        {
            "tracking_source":
                "CTC reference",

            "n_three_step_sequences":
                len(
                    ctc_three_step_raw
                ),

            "n_three_step_groups":
                ctc_three_step_raw[
                    "cell_id"
                ].nunique(),

            "n_four_step_sequences":
                len(
                    ctc_four_step_raw
                ),

            "n_four_step_groups":
                ctc_four_step_raw[
                    "cell_id"
                ].nunique(),
        },
        {
            "tracking_source":
                "TrackMate",

            "n_three_step_sequences":
                len(
                    trackmate_three_step_raw
                ),

            "n_three_step_groups":
                trackmate_three_step_raw[
                    "cell_id"
                ].nunique(),

            "n_four_step_sequences":
                len(
                    trackmate_four_step_raw
                ),

            "n_four_step_groups":
                trackmate_four_step_raw[
                    "cell_id"
                ].nunique(),
        },
    ]
)


support_by_source[
    "four_step_eligible"
] = (
    (
        support_by_source[
            "n_four_step_sequences"
        ]
        >= FOUR_STEP_MIN_SEQUENCES
    )
    &
    (
        support_by_source[
            "n_four_step_groups"
        ]
        >= FOUR_STEP_MIN_GROUPS
    )
)


support_audit_records = []

for _, row in (
    support_by_source.iterrows()
):

    for representation in [
        "raw",
        "co-moving",
    ]:

        support_audit_records.append(
            {
                "tracking_source":
                    row[
                        "tracking_source"
                    ],

                "representation":
                    representation,

                "n_three_step_sequences":
                    int(
                        row[
                            "n_three_step_sequences"
                        ]
                    ),

                "n_three_step_groups":
                    int(
                        row[
                            "n_three_step_groups"
                        ]
                    ),

                "n_four_step_sequences":
                    int(
                        row[
                            "n_four_step_sequences"
                        ]
                    ),

                "n_four_step_groups":
                    int(
                        row[
                            "n_four_step_groups"
                        ]
                    ),

                "four_step_eligible":
                    bool(
                        row[
                            "four_step_eligible"
                        ]
                    ),
            }
        )


phase8b_support_audit = (
    pd.DataFrame(
        support_audit_records
    )
)


display(
    phase8b_support_audit
)


print(
    "\nFour-step gate:"
)

print(
    "minimum sequences =",
    FOUR_STEP_MIN_SEQUENCES,
)

print(
    "minimum groups =",
    FOUR_STEP_MIN_GROUPS,
)


print(
    "\nPhase 8B sample-support audit: COMPLETE"
)

,tracking_source,representation,n_three_step_sequences,n_three_step_groups,n_four_step_sequences,n_four_step_groups,four_step_eligible
0,CTC reference,raw,383,14,369,14,True
1,CTC reference,co-moving,383,14,369,14,True
2,TrackMate,raw,357,28,329,26,True
3,TrackMate,co-moving,357,28,329,26,True



Four-step gate:
minimum sequences = 200
minimum groups = 12

Phase 8B sample-support audit: COMPLETE


## Phase 8B.1 — Exact sequence-support audit

The first data-dependent Phase 8B computation was restricted to temporal
sample support, as required by the pre-specified analysis design.

No classifier performance, null statistic, p-value, or irreversibility
quantity was inspected.

### Three-step support

CTC reference tracking produced:

- 383 exact three-step sequences;
- 14 contributing reference tracks.

TrackMate produced:

- 357 exact three-step sequences;
- 28 contributing tracks.

### Four-step support

CTC reference tracking produced:

- 369 exact four-step sequences;
- 14 contributing reference tracks.

TrackMate produced:

- 329 exact four-step sequences;
- 26 contributing tracks.

The pre-specified four-step eligibility rule was:

`at least 200 exact four-step sequences AND at least 12 trajectory groups`.

Both tracking sources satisfy this rule.

Therefore:

- three-step analysis remains the primary Phase 8B analysis;
- four-step analysis is formally eligible and will be retained as the
  pre-specified secondary extension;
- the four-step support thresholds will not be modified later in response to
  inferential results.

Raw and co-moving representations have identical temporal support because
common-motion subtraction changes trajectory coordinates but does not add or
remove track-frame observations.

The support audit is now closed.

In [7]:
ctc_four_step_eligible = bool(
    support_by_source.loc[
        support_by_source[
            "tracking_source"
        ].eq("CTC reference"),
        "four_step_eligible",
    ].iloc[0]
)

trackmate_four_step_eligible = bool(
    support_by_source.loc[
        support_by_source[
            "tracking_source"
        ].eq("TrackMate"),
        "four_step_eligible",
    ].iloc[0]
)


assert ctc_four_step_eligible
assert trackmate_four_step_eligible


print(
    "CTC four-step eligibility:",
    ctc_four_step_eligible,
)

print(
    "TrackMate four-step eligibility:",
    trackmate_four_step_eligible,
)

print(
    "Phase 8B support-gate decision: RECORDED"
)

CTC four-step eligibility: True
TrackMate four-step eligibility: True
Phase 8B support-gate decision: RECORDED


## Phase 8B.2 — Three-step reversal and parity validation

The three-step reversal and parity-coordinate definitions were implemented and
protected by synthetic unit tests before being applied to the real trajectory
data.

For

\[
S_3=(\mathbf d_1,\mathbf d_2,\mathbf d_3),
\]

the exact reversed sequence is

\[
R(S_3)=(-\mathbf d_3,-\mathbf d_2,-\mathbf d_1).
\]

The pre-specified parity coordinates are

\[
\mathbf n=\mathbf d_1+\mathbf d_2+\mathbf d_3,
\]

\[
\mathbf q=\mathbf d_1-2\mathbf d_2+\mathbf d_3,
\]

and

\[
\mathbf e=\mathbf d_1-\mathbf d_3.
\]

Under exact reversal,

\[
\mathbf n\rightarrow-\mathbf n,
\qquad
\mathbf q\rightarrow-\mathbf q,
\qquad
\mathbf e\rightarrow\mathbf e.
\]

The transformation is invertible, so the full six-dimensional three-step
sequence can be reconstructed exactly from \((\mathbf n,\mathbf q,\mathbf e)\).

This section performs only structural and algebraic validation on the real
data. No classifier performance or inferential statistic is evaluated yet.

In [8]:
from src.sequences import (
    get_k_step_sequence_array,
    reverse_k_step_sequences,
    decompose_three_step_parity,
    reconstruct_three_step_from_parity,
)

In [9]:
ctc_three_step_raw_array = (
    get_k_step_sequence_array(
        ctc_three_step_raw,
        n_steps=3,
    )
)

trackmate_three_step_raw_array = (
    get_k_step_sequence_array(
        trackmate_three_step_raw,
        n_steps=3,
    )
)


print(
    "CTC raw 3-step array shape:",
    ctc_three_step_raw_array.shape,
)

print(
    "TrackMate raw 3-step array shape:",
    trackmate_three_step_raw_array.shape,
)


assert (
    ctc_three_step_raw_array.shape
    == (383, 6)
)

assert (
    trackmate_three_step_raw_array.shape
    == (357, 6)
)


print(
    "Real raw three-step arrays: PASSED"
)

CTC raw 3-step array shape: (383, 6)
TrackMate raw 3-step array shape: (357, 6)
Real raw three-step arrays: PASSED


In [10]:
ctc_three_step_raw_reverse = (
    reverse_k_step_sequences(
        ctc_three_step_raw_array
    )
)

trackmate_three_step_raw_reverse = (
    reverse_k_step_sequences(
        trackmate_three_step_raw_array
    )
)


assert np.allclose(
    reverse_k_step_sequences(
        ctc_three_step_raw_reverse
    ),
    ctc_three_step_raw_array,
)

assert np.allclose(
    reverse_k_step_sequences(
        trackmate_three_step_raw_reverse
    ),
    trackmate_three_step_raw_array,
)


print(
    "CTC reversal involution: PASSED"
)

print(
    "TrackMate reversal involution: PASSED"
)

CTC reversal involution: PASSED
TrackMate reversal involution: PASSED


In [11]:
(
    ctc_raw_net_odd,
    ctc_raw_internal_odd,
    ctc_raw_even,
) = decompose_three_step_parity(
    ctc_three_step_raw_array
)


(
    trackmate_raw_net_odd,
    trackmate_raw_internal_odd,
    trackmate_raw_even,
) = decompose_three_step_parity(
    trackmate_three_step_raw_array
)


print(
    "CTC parity shapes:",
    ctc_raw_net_odd.shape,
    ctc_raw_internal_odd.shape,
    ctc_raw_even.shape,
)

print(
    "TrackMate parity shapes:",
    trackmate_raw_net_odd.shape,
    trackmate_raw_internal_odd.shape,
    trackmate_raw_even.shape,
)


assert (
    ctc_raw_net_odd.shape
    == (383, 2)
)

assert (
    trackmate_raw_net_odd.shape
    == (357, 2)
)


print(
    "Real raw three-step parity decomposition: PASSED"
)

CTC parity shapes: (383, 2) (383, 2) (383, 2)
TrackMate parity shapes: (357, 2) (357, 2) (357, 2)
Real raw three-step parity decomposition: PASSED


In [12]:
(
    ctc_raw_reverse_net_odd,
    ctc_raw_reverse_internal_odd,
    ctc_raw_reverse_even,
) = decompose_three_step_parity(
    ctc_three_step_raw_reverse
)


(
    trackmate_raw_reverse_net_odd,
    trackmate_raw_reverse_internal_odd,
    trackmate_raw_reverse_even,
) = decompose_three_step_parity(
    trackmate_three_step_raw_reverse
)


# CTC
assert np.allclose(
    ctc_raw_reverse_net_odd,
    -ctc_raw_net_odd,
)

assert np.allclose(
    ctc_raw_reverse_internal_odd,
    -ctc_raw_internal_odd,
)

assert np.allclose(
    ctc_raw_reverse_even,
    ctc_raw_even,
)


# TrackMate
assert np.allclose(
    trackmate_raw_reverse_net_odd,
    -trackmate_raw_net_odd,
)

assert np.allclose(
    trackmate_raw_reverse_internal_odd,
    -trackmate_raw_internal_odd,
)

assert np.allclose(
    trackmate_raw_reverse_even,
    trackmate_raw_even,
)


print(
    "Real-data three-step reversal-parity algebra: PASSED"
)

Real-data three-step reversal-parity algebra: PASSED


In [13]:
ctc_raw_reconstructed = (
    reconstruct_three_step_from_parity(
        ctc_raw_net_odd,
        ctc_raw_internal_odd,
        ctc_raw_even,
    )
)


trackmate_raw_reconstructed = (
    reconstruct_three_step_from_parity(
        trackmate_raw_net_odd,
        trackmate_raw_internal_odd,
        trackmate_raw_even,
    )
)


ctc_roundtrip_max_error = float(
    np.max(
        np.abs(
            ctc_raw_reconstructed
            - ctc_three_step_raw_array
        )
    )
)

trackmate_roundtrip_max_error = float(
    np.max(
        np.abs(
            trackmate_raw_reconstructed
            - trackmate_three_step_raw_array
        )
    )
)


print(
    "CTC maximum parity round-trip error:",
    ctc_roundtrip_max_error,
)

print(
    "TrackMate maximum parity round-trip error:",
    trackmate_roundtrip_max_error,
)


assert np.allclose(
    ctc_raw_reconstructed,
    ctc_three_step_raw_array,
)

assert np.allclose(
    trackmate_raw_reconstructed,
    trackmate_three_step_raw_array,
)


print(
    "Real-data parity reconstruction: PASSED"
)

CTC maximum parity round-trip error: 3.552713678800501e-15
TrackMate maximum parity round-trip error: 1.7763568394002505e-15
Real-data parity reconstruction: PASSED


## Reconstructing the co-moving representation

TrackMate already has a historically validated saved co-moving trajectory
table.

CTC reference tracking does not require a separate committed co-moving data
file.

Instead, Phase 8B will reconstruct co-moving coordinates using the same
frame-wise common-motion definition used previously.

Before applying that reconstruction to CTC, the transformation will be
regression-checked against the saved TrackMate co-moving trajectory table.

The required validation is:

> applying the reconstructed common-motion transformation to raw TrackMate
> trajectories must reproduce the historically saved TrackMate co-moving
> coordinates.

Only after this check passes will the same transformation be applied to the
CTC reference trajectories.

This prevents an accidental change in the definition of the co-moving
representation during Phase 8B.

In [14]:
from src.steps import compute_steps_for_tau


def compute_framewise_common_motion(
    tracks,
):
    """
    Estimate frame-wise population common motion
    from exact one-frame displacements.

    For each frame transition, common dx and dy
    are the component-wise medians across cells.
    """

    one_frame_steps = (
        compute_steps_for_tau(
            tracks,
            tau_frames=1,
        )
    )

    common_motion = (
        one_frame_steps
        .groupby(
            [
                "frame_start",
                "frame_end",
            ],
            as_index=False,
        )
        .agg(
            common_dx_um=(
                "dx_um",
                "median",
            ),
            common_dy_um=(
                "dy_um",
                "median",
            ),
            n_cells=(
                "cell_id",
                "nunique",
            ),
        )
        .sort_values(
            "frame_start"
        )
        .reset_index(
            drop=True
        )
    )

    return common_motion

In [15]:
def make_comoving_tracks(
    tracks,
    common_motion,
):
    """
    Subtract cumulative frame-wise population
    common translation from trajectory coordinates.

    Frame 0 has zero cumulative correction.
    """

    frame_correction = pd.DataFrame(
        {
            "frame": [0],
            "common_x_um": [0.0],
            "common_y_um": [0.0],
        }
    )

    transition_correction = pd.DataFrame(
        {
            "frame":
                common_motion[
                    "frame_end"
                ].to_numpy(),

            "common_x_um":
                np.cumsum(
                    common_motion[
                        "common_dx_um"
                    ].to_numpy()
                ),

            "common_y_um":
                np.cumsum(
                    common_motion[
                        "common_dy_um"
                    ].to_numpy()
                ),
        }
    )

    frame_correction = pd.concat(
        [
            frame_correction,
            transition_correction,
        ],
        ignore_index=True,
    )

    result = (
        tracks
        .merge(
            frame_correction,
            on="frame",
            how="left",
            validate="many_to_one",
        )
        .copy()
    )

    if (
        result[
            [
                "common_x_um",
                "common_y_um",
            ]
        ]
        .isna()
        .any()
        .any()
    ):
        raise RuntimeError(
            "Missing cumulative common-motion "
            "correction for one or more frames."
        )

    result[
        "x_um"
    ] = (
        result["x_um"]
        - result["common_x_um"]
    )

    result[
        "y_um"
    ] = (
        result["y_um"]
        - result["common_y_um"]
    )

    return (
        result[
            [
                "cell_id",
                "frame",
                "t_min",
                "x_um",
                "y_um",
            ]
        ]
        .sort_values(
            [
                "cell_id",
                "frame",
            ]
        )
        .reset_index(
            drop=True
        )
    )

In [16]:
trackmate_common_motion_phase8b = (
    compute_framewise_common_motion(
        trackmate_raw
    )
)


trackmate_comoving_reconstructed = (
    make_comoving_tracks(
        trackmate_raw,
        trackmate_common_motion_phase8b,
    )
)


trackmate_comoving_expected = (
    trackmate_comoving[
        [
            "cell_id",
            "frame",
            "t_min",
            "x_um",
            "y_um",
        ]
    ]
    .sort_values(
        [
            "cell_id",
            "frame",
        ]
    )
    .reset_index(
        drop=True
    )
)


print(
    "TrackMate common-motion transitions:",
    len(
        trackmate_common_motion_phase8b
    ),
)


print(
    "Maximum x reconstruction difference:",
    np.max(
        np.abs(
            trackmate_comoving_reconstructed[
                "x_um"
            ].to_numpy()
            - trackmate_comoving_expected[
                "x_um"
            ].to_numpy()
        )
    ),
)

print(
    "Maximum y reconstruction difference:",
    np.max(
        np.abs(
            trackmate_comoving_reconstructed[
                "y_um"
            ].to_numpy()
            - trackmate_comoving_expected[
                "y_um"
            ].to_numpy()
        )
    ),
)


assert np.array_equal(
    trackmate_comoving_reconstructed[
        [
            "cell_id",
            "frame",
        ]
    ].to_numpy(),
    trackmate_comoving_expected[
        [
            "cell_id",
            "frame",
        ]
    ].to_numpy(),
)


assert np.allclose(
    trackmate_comoving_reconstructed[
        [
            "x_um",
            "y_um",
        ]
    ].to_numpy(),
    trackmate_comoving_expected[
        [
            "x_um",
            "y_um",
        ]
    ].to_numpy(),
    atol=1e-10,
)


print(
    "TrackMate co-moving reconstruction regression: PASSED"
)

TrackMate common-motion transitions: 47
Maximum x reconstruction difference: 5.684341886080802e-14
Maximum y reconstruction difference: 2.842170943040401e-14
TrackMate co-moving reconstruction regression: PASSED


In [17]:
ctc_common_motion_phase8b = (
    compute_framewise_common_motion(
        ctc_raw
    )
)


ctc_comoving = (
    make_comoving_tracks(
        ctc_raw,
        ctc_common_motion_phase8b,
    )
)


validate_tracks(
    ctc_comoving
)


assert ctc_comoving[
    [
        "cell_id",
        "frame",
        "t_min",
    ]
].equals(
    ctc_raw[
        [
            "cell_id",
            "frame",
            "t_min",
        ]
    ]
)


print(
    "CTC common-motion transitions:",
    len(
        ctc_common_motion_phase8b
    ),
)

print(
    "CTC co-moving observations:",
    len(ctc_comoving),
)

print(
    "CTC co-moving tracks:",
    ctc_comoving[
        "cell_id"
    ].nunique(),
)

print(
    "CTC co-moving reconstruction: PASSED"
)

CTC common-motion transitions: 47
CTC co-moving observations: 428
CTC co-moving tracks: 15
CTC co-moving reconstruction: PASSED


In [18]:
ctc_three_step_comoving = (
    build_k_step_sequences(
        ctc_comoving,
        n_steps=3,
    )
)

trackmate_three_step_comoving = (
    build_k_step_sequences(
        trackmate_comoving,
        n_steps=3,
    )
)


assert len(
    ctc_three_step_comoving
) == 383

assert len(
    trackmate_three_step_comoving
) == 357


ctc_window_identity = np.array_equal(
    ctc_three_step_raw[
        [
            "cell_id",
            "frame_start",
            "frame_end",
        ]
    ].to_numpy(),
    ctc_three_step_comoving[
        [
            "cell_id",
            "frame_start",
            "frame_end",
        ]
    ].to_numpy(),
)


trackmate_window_identity = (
    np.array_equal(
        trackmate_three_step_raw[
            [
                "cell_id",
                "frame_start",
                "frame_end",
            ]
        ].to_numpy(),
        trackmate_three_step_comoving[
            [
                "cell_id",
                "frame_start",
                "frame_end",
            ]
        ].to_numpy(),
    )
)


print(
    "CTC raw/co-moving window identity:",
    ctc_window_identity,
)

print(
    "TrackMate raw/co-moving window identity:",
    trackmate_window_identity,
)


assert ctc_window_identity
assert trackmate_window_identity


print(
    "Three-step raw/co-moving sequence identity: PASSED"
)

CTC raw/co-moving window identity: True
TrackMate raw/co-moving window identity: True
Three-step raw/co-moving sequence identity: PASSED


## Phase 8B.3 — Pre-classification feature-space and fold audit

The real-data three-step reversal and parity algebra has now been validated in
both tracking sources.

Before inspecting any classifier performance, the exact feature
representations and cross-validation folds are fixed.

The five pre-specified three-step feature spaces are:

### 1. `net_only`

$$
\mathbf{n}
=
\mathbf{d}_1+\mathbf{d}_2+\mathbf{d}_3
$$

This is the net displacement across the complete three-step interval.

---

### 2. `internal_odd_only`

$$
\mathbf{q}
=
\mathbf{d}_1
-
2\mathbf{d}_2
+
\mathbf{d}_3
$$

This reversal-odd coordinate depends on the internal distribution of motion
across the three ordered steps.

---

### 3. `even_only`

$$
\mathbf{e}
=
\mathbf{d}_1-\mathbf{d}_3
$$

This coordinate is invariant under exact time reversal and therefore serves as
a negative-control feature space.

---

### 4. `odd_combined`

$$
[\mathbf{n},\mathbf{q}]
$$

This four-dimensional representation contains both the net directional
component and the internal reversal-odd component.

---

### 5. `full_6d`

$$
[
\mathbf{d}_1,
\mathbf{d}_2,
\mathbf{d}_3
]
$$

This is the original physical six-dimensional displacement representation:

`(dx1, dy1, dx2, dy2, dx3, dy3)`

The full feature space will use the original displacement coordinates rather
than the equivalent parity-coordinate representation.

Although the two full representations are related by an invertible linear
transformation and therefore contain the same information, a regularized
classifier combined with feature scaling need not produce numerically
identical fitted models under an arbitrary change of coordinates.

Therefore, the original displacement coordinates are fixed here, before any
Phase 8B classifier AUC is inspected.

The cross-validation design remains the pre-specified `3-fold GroupKFold`,
with trajectory identity as the grouping variable.

No classifier is fitted in this section.

In [19]:
ctc_three_step_comoving_array = (
    get_k_step_sequence_array(
        ctc_three_step_comoving,
        n_steps=3,
    )
)


trackmate_three_step_comoving_array = (
    get_k_step_sequence_array(
        trackmate_three_step_comoving,
        n_steps=3,
    )
)


assert (
    ctc_three_step_comoving_array.shape
    == (383, 6)
)

assert (
    trackmate_three_step_comoving_array.shape
    == (357, 6)
)


print(
    "CTC co-moving 3-step array shape:",
    ctc_three_step_comoving_array.shape,
)

print(
    "TrackMate co-moving 3-step array shape:",
    trackmate_three_step_comoving_array.shape,
)

print(
    "Real co-moving three-step arrays: PASSED"
)

CTC co-moving 3-step array shape: (383, 6)
TrackMate co-moving 3-step array shape: (357, 6)
Real co-moving three-step arrays: PASSED


In [20]:
ctc_three_step_comoving_reverse = (
    reverse_k_step_sequences(
        ctc_three_step_comoving_array
    )
)


trackmate_three_step_comoving_reverse = (
    reverse_k_step_sequences(
        trackmate_three_step_comoving_array
    )
)


(
    ctc_comoving_net_odd,
    ctc_comoving_internal_odd,
    ctc_comoving_even,
) = decompose_three_step_parity(
    ctc_three_step_comoving_array
)


(
    ctc_comoving_reverse_net_odd,
    ctc_comoving_reverse_internal_odd,
    ctc_comoving_reverse_even,
) = decompose_three_step_parity(
    ctc_three_step_comoving_reverse
)


(
    trackmate_comoving_net_odd,
    trackmate_comoving_internal_odd,
    trackmate_comoving_even,
) = decompose_three_step_parity(
    trackmate_three_step_comoving_array
)


(
    trackmate_comoving_reverse_net_odd,
    trackmate_comoving_reverse_internal_odd,
    trackmate_comoving_reverse_even,
) = decompose_three_step_parity(
    trackmate_three_step_comoving_reverse
)


# Exact reversal involution.
assert np.allclose(
    reverse_k_step_sequences(
        ctc_three_step_comoving_reverse
    ),
    ctc_three_step_comoving_array,
)

assert np.allclose(
    reverse_k_step_sequences(
        trackmate_three_step_comoving_reverse
    ),
    trackmate_three_step_comoving_array,
)


# CTC parity transformation.
assert np.allclose(
    ctc_comoving_reverse_net_odd,
    -ctc_comoving_net_odd,
)

assert np.allclose(
    ctc_comoving_reverse_internal_odd,
    -ctc_comoving_internal_odd,
)

assert np.allclose(
    ctc_comoving_reverse_even,
    ctc_comoving_even,
)


# TrackMate parity transformation.
assert np.allclose(
    trackmate_comoving_reverse_net_odd,
    -trackmate_comoving_net_odd,
)

assert np.allclose(
    trackmate_comoving_reverse_internal_odd,
    -trackmate_comoving_internal_odd,
)

assert np.allclose(
    trackmate_comoving_reverse_even,
    trackmate_comoving_even,
)


print(
    "Real co-moving three-step reversal-parity algebra: PASSED"
)

Real co-moving three-step reversal-parity algebra: PASSED


In [21]:
ctc_comoving_reconstructed = (
    reconstruct_three_step_from_parity(
        ctc_comoving_net_odd,
        ctc_comoving_internal_odd,
        ctc_comoving_even,
    )
)


trackmate_comoving_reconstructed = (
    reconstruct_three_step_from_parity(
        trackmate_comoving_net_odd,
        trackmate_comoving_internal_odd,
        trackmate_comoving_even,
    )
)


ctc_comoving_roundtrip_max_error = float(
    np.max(
        np.abs(
            ctc_comoving_reconstructed
            - ctc_three_step_comoving_array
        )
    )
)


trackmate_comoving_roundtrip_max_error = float(
    np.max(
        np.abs(
            trackmate_comoving_reconstructed
            - trackmate_three_step_comoving_array
        )
    )
)


print(
    "CTC co-moving maximum parity round-trip error:",
    ctc_comoving_roundtrip_max_error,
)

print(
    "TrackMate co-moving maximum parity round-trip error:",
    trackmate_comoving_roundtrip_max_error,
)


assert np.allclose(
    ctc_comoving_reconstructed,
    ctc_three_step_comoving_array,
)

assert np.allclose(
    trackmate_comoving_reconstructed,
    trackmate_three_step_comoving_array,
)


print(
    "Co-moving parity reconstruction: PASSED"
)

CTC co-moving maximum parity round-trip error: 3.552713678800501e-15
TrackMate co-moving maximum parity round-trip error: 1.7763568394002505e-15
Co-moving parity reconstruction: PASSED


In [22]:
def make_three_step_feature_spaces(
    forward_array,
):
    """
    Construct the five pre-specified Phase 8B
    three-step feature spaces and their exact
    reversed counterparts.
    """

    reverse_array = (
        reverse_k_step_sequences(
            forward_array
        )
    )

    (
        net_forward,
        internal_forward,
        even_forward,
    ) = decompose_three_step_parity(
        forward_array
    )

    (
        net_reverse,
        internal_reverse,
        even_reverse,
    ) = decompose_three_step_parity(
        reverse_array
    )

    forward_spaces = {
        "net_only":
            net_forward,

        "internal_odd_only":
            internal_forward,

        "even_only":
            even_forward,

        "odd_combined":
            np.column_stack(
                [
                    net_forward,
                    internal_forward,
                ]
            ),

        "full_6d":
            forward_array,
    }

    reverse_spaces = {
        "net_only":
            net_reverse,

        "internal_odd_only":
            internal_reverse,

        "even_only":
            even_reverse,

        "odd_combined":
            np.column_stack(
                [
                    net_reverse,
                    internal_reverse,
                ]
            ),

        "full_6d":
            reverse_array,
    }

    return (
        forward_spaces,
        reverse_spaces,
    )

In [23]:
(
    ctc_raw_feature_spaces,
    ctc_raw_reverse_feature_spaces,
) = make_three_step_feature_spaces(
    ctc_three_step_raw_array
)


(
    ctc_comoving_feature_spaces,
    ctc_comoving_reverse_feature_spaces,
) = make_three_step_feature_spaces(
    ctc_three_step_comoving_array
)


(
    trackmate_raw_feature_spaces,
    trackmate_raw_reverse_feature_spaces,
) = make_three_step_feature_spaces(
    trackmate_three_step_raw_array
)


(
    trackmate_comoving_feature_spaces,
    trackmate_comoving_reverse_feature_spaces,
) = make_three_step_feature_spaces(
    trackmate_three_step_comoving_array
)


expected_feature_dimensions = {
    "net_only": 2,
    "internal_odd_only": 2,
    "even_only": 2,
    "odd_combined": 4,
    "full_6d": 6,
}


for (
    source_label,
    feature_spaces,
    expected_rows,
) in [
    (
        "CTC raw",
        ctc_raw_feature_spaces,
        383,
    ),
    (
        "CTC co-moving",
        ctc_comoving_feature_spaces,
        383,
    ),
    (
        "TrackMate raw",
        trackmate_raw_feature_spaces,
        357,
    ),
    (
        "TrackMate co-moving",
        trackmate_comoving_feature_spaces,
        357,
    ),
]:

    for (
        feature_name,
        expected_dimension,
    ) in expected_feature_dimensions.items():

        observed_shape = (
            feature_spaces[
                feature_name
            ].shape
        )

        assert observed_shape == (
            expected_rows,
            expected_dimension,
        )


print(
    "All pre-specified three-step feature spaces: PASSED"
)

All pre-specified three-step feature spaces: PASSED


In [24]:
for (
    source_label,
    forward_spaces,
    reverse_spaces,
) in [
    (
        "CTC raw",
        ctc_raw_feature_spaces,
        ctc_raw_reverse_feature_spaces,
    ),
    (
        "CTC co-moving",
        ctc_comoving_feature_spaces,
        ctc_comoving_reverse_feature_spaces,
    ),
    (
        "TrackMate raw",
        trackmate_raw_feature_spaces,
        trackmate_raw_reverse_feature_spaces,
    ),
    (
        "TrackMate co-moving",
        trackmate_comoving_feature_spaces,
        trackmate_comoving_reverse_feature_spaces,
    ),
]:

    assert np.allclose(
        forward_spaces[
            "even_only"
        ],
        reverse_spaces[
            "even_only"
        ],
    )

    assert np.allclose(
        forward_spaces[
            "net_only"
        ],
        -reverse_spaces[
            "net_only"
        ],
    )

    assert np.allclose(
        forward_spaces[
            "internal_odd_only"
        ],
        -reverse_spaces[
            "internal_odd_only"
        ],
    )


print(
    "Feature-space reversal sanity checks: PASSED"
)

Feature-space reversal sanity checks: PASSED


In [25]:
from src.classification import (
    build_classification_dataset,
)


ctc_three_step_groups = (
    ctc_three_step_raw[
        "cell_id"
    ].to_numpy()
)


trackmate_three_step_groups = (
    trackmate_three_step_raw[
        "cell_id"
    ].to_numpy()
)


def build_feature_classification_sets(
    forward_spaces,
    reverse_spaces,
    groups,
):
    """
    Build balanced forward/reversed classifier
    datasets for every pre-specified feature space.
    """

    datasets = {}

    for feature_name in (
        forward_spaces.keys()
    ):

        (
            X,
            y,
            classifier_groups,
        ) = build_classification_dataset(
            forward_spaces[
                feature_name
            ],
            reverse_spaces[
                feature_name
            ],
            groups,
        )

        datasets[
            feature_name
        ] = {
            "X": X,
            "y": y,
            "groups":
                classifier_groups,
        }

    return datasets

In [26]:
ctc_raw_classifier_sets = (
    build_feature_classification_sets(
        ctc_raw_feature_spaces,
        ctc_raw_reverse_feature_spaces,
        ctc_three_step_groups,
    )
)


ctc_comoving_classifier_sets = (
    build_feature_classification_sets(
        ctc_comoving_feature_spaces,
        ctc_comoving_reverse_feature_spaces,
        ctc_three_step_groups,
    )
)


trackmate_raw_classifier_sets = (
    build_feature_classification_sets(
        trackmate_raw_feature_spaces,
        trackmate_raw_reverse_feature_spaces,
        trackmate_three_step_groups,
    )
)


trackmate_comoving_classifier_sets = (
    build_feature_classification_sets(
        trackmate_comoving_feature_spaces,
        trackmate_comoving_reverse_feature_spaces,
        trackmate_three_step_groups,
    )
)

In [27]:
for (
    source_label,
    classifier_sets,
    expected_examples,
    expected_groups,
) in [
    (
        "CTC raw",
        ctc_raw_classifier_sets,
        766,
        14,
    ),
    (
        "CTC co-moving",
        ctc_comoving_classifier_sets,
        766,
        14,
    ),
    (
        "TrackMate raw",
        trackmate_raw_classifier_sets,
        714,
        28,
    ),
    (
        "TrackMate co-moving",
        trackmate_comoving_classifier_sets,
        714,
        28,
    ),
]:

    for (
        feature_name,
        dataset,
    ) in classifier_sets.items():

        X = dataset["X"]
        y = dataset["y"]
        groups = dataset["groups"]

        assert len(X) == expected_examples
        assert len(y) == expected_examples
        assert len(groups) == expected_examples

        assert np.isclose(
            y.mean(),
            0.5,
        )

        assert (
            np.unique(groups).size
            == expected_groups
        )


print(
    "Three-step classifier dataset structure: PASSED"
)

Three-step classifier dataset structure: PASSED


In [28]:
from sklearn.model_selection import (
    GroupKFold,
)


PHASE8B_N_SPLITS = 3


def audit_group_folds(
    X,
    y,
    groups,
    cv,
):
    """
    Audit grouped cross-validation structure
    without fitting a classifier.
    """

    records = []

    for (
        fold_id,
        (
            train_indices,
            test_indices,
        ),
    ) in enumerate(
        cv.split(
            X,
            y,
            groups,
        ),
        start=1,
    ):

        train_groups = np.unique(
            groups[
                train_indices
            ]
        )

        test_groups = np.unique(
            groups[
                test_indices
            ]
        )

        overlap = np.intersect1d(
            train_groups,
            test_groups,
        )

        records.append(
            {
                "fold":
                    fold_id,

                "n_train_examples":
                    len(
                        train_indices
                    ),

                "n_test_examples":
                    len(
                        test_indices
                    ),

                "n_train_groups":
                    len(
                        train_groups
                    ),

                "n_test_groups":
                    len(
                        test_groups
                    ),

                "train_label_fraction":
                    float(
                        y[
                            train_indices
                        ].mean()
                    ),

                "test_label_fraction":
                    float(
                        y[
                            test_indices
                        ].mean()
                    ),

                "group_overlap":
                    len(
                        overlap
                    ),
            }
        )

    return pd.DataFrame(
        records
    )

In [29]:
ctc_phase8b_cv = GroupKFold(
    n_splits=PHASE8B_N_SPLITS
)

trackmate_phase8b_cv = GroupKFold(
    n_splits=PHASE8B_N_SPLITS
)


ctc_fold_audit = (
    audit_group_folds(
        X=ctc_raw_classifier_sets[
            "net_only"
        ]["X"],
        y=ctc_raw_classifier_sets[
            "net_only"
        ]["y"],
        groups=ctc_raw_classifier_sets[
            "net_only"
        ]["groups"],
        cv=ctc_phase8b_cv,
    )
)


trackmate_fold_audit = (
    audit_group_folds(
        X=trackmate_raw_classifier_sets[
            "net_only"
        ]["X"],
        y=trackmate_raw_classifier_sets[
            "net_only"
        ]["y"],
        groups=trackmate_raw_classifier_sets[
            "net_only"
        ]["groups"],
        cv=trackmate_phase8b_cv,
    )
)


print(
    "CTC 3-fold audit:"
)

display(
    ctc_fold_audit
)


print(
    "TrackMate 3-fold audit:"
)

display(
    trackmate_fold_audit
)

CTC 3-fold audit:


,fold,n_train_examples,n_test_examples,n_train_groups,n_test_groups,train_label_fraction,test_label_fraction,group_overlap
0,1,516,250,10,4,0.5,0.5,0
1,2,514,252,9,5,0.5,0.5,0
2,3,502,264,9,5,0.5,0.5,0


TrackMate 3-fold audit:


,fold,n_train_examples,n_test_examples,n_train_groups,n_test_groups,train_label_fraction,test_label_fraction,group_overlap
0,1,476,238,19,9,0.5,0.5,0
1,2,476,238,19,9,0.5,0.5,0
2,3,476,238,18,10,0.5,0.5,0


In [30]:
for (
    source_label,
    fold_audit,
) in [
    (
        "CTC",
        ctc_fold_audit,
    ),
    (
        "TrackMate",
        trackmate_fold_audit,
    ),
]:

    assert (
        fold_audit[
            "group_overlap"
        ]
        == 0
    ).all()

    assert np.allclose(
        fold_audit[
            "train_label_fraction"
        ],
        0.5,
    )

    assert np.allclose(
        fold_audit[
            "test_label_fraction"
        ],
        0.5,
    )

    assert (
        fold_audit[
            "n_test_groups"
        ]
        >= 3
    ).all()


print(
    "Phase 8B grouped-fold validation: PASSED"
)

Phase 8B grouped-fold validation: PASSED


### Pre-classification checkpoint

The complete three-step feature and validation structure is now fixed before
classifier performance is inspected.

For each tracking source and each raw/co-moving representation, five
pre-specified feature spaces have been constructed:

- net displacement only;
- internal reversal-odd structure only;
- reversal-even control;
- combined reversal-odd coordinates;
- full original six-dimensional sequence.

Forward and reversed datasets are exactly balanced.

All examples from a trajectory remain within the same cross-validation fold.

Three-fold GroupKFold produces no train/test trajectory overlap and preserves
the exact 50/50 forward/reversed class balance.

The even-only forward and reversed feature arrays are numerically identical,
providing a strong negative-control representation.

No classifier has yet been fitted and no Phase 8B AUC has yet been inspected.